### PR TITLE
Note Improvements, text editing and bullet point support

### PR DESCRIPTION
**Pull request by martinstacey**
### Purpose

The purpose of this PR is to simplify the `Note` text editing UI. It allows the user to double click the note and edit it directly. Similar to how groups/annotations work. See the image below for reference:

noteEditingFromTextBox

It also extends the functionality of the note by adding bullet point support.
It changes dashes to bullets and also handles indentation for different bullets as shown in the image below:

indented new

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@Amoursol 
@QilongTang 

### FYIs
 @SHKnudsen 
@M-JULIANI

### CODE REVIEW

**Comment by Amoursol** [comment]
Hello @martinstacey - This is looking awesome! 🚀  So much improvement to the overall experience for our users here. 

Is it possible to change the depth of the indented bullet points? They are currently quite big - something around 40% of what it is right now would be awesome.

**Comment by martinstacey** [comment]
> Hello @martinstacey - This is looking awesome! 🚀 So much improvement to the overall experience for our users here.
> 
> Is it possible to change the depth of the indented bullet points? They are currently quite big - something around 40% of what it is right now would be awesome.

Thanks!
the tab spacing has been driving me crazy 😂 as there is no direct way of changing the tab spacing in textboxs without converting them to spaces, but this may cause unexpected behavior on how the text is read as indentation. Let me give this a think 🤔 
https://social.msdn.microsoft.com/Forums/vstudio/en-US/00c1ea7f-0a2c-4290-bdc3-26685c77f5aa/tab-spacing-in-textbox-when-using-tab-key?forum=wpf

**Comment by martinstacey** [comment]
@Amoursol, ok solved the issue on indentation
bulletIndentation

**Comment by Amoursol** [comment]
> @Amoursol, ok solved the issue on indentation
> bulletIndentation

Brilliant - looks much better indeed @martinstacey !

**Comment by martinstacey** [comment]
Hi @Amoursol, @QilongTang  to confirm on both things pending on this PR, editing using `Edit...` both far and close.
Adding space after bullet

image

image

**Comment by QilongTang** [comment]
Thanks @martinstacey I will give this some review soon!

File: src/DynamoCoreWpf/UI/Prompts/EditWindow.xaml.cs [hunk 1]
```diff
@@ -40,7 +47,14 @@ public partial class EditWindow
                         expr.UpdateSource();
                 };
             }
+            this.editText.PreviewKeyDown += EditText_PreviewKeyDown;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/UI/Prompts/EditWindow.xaml.cs:50]: Should we unsubscribe this somewhere?

File: src/DynamoCoreWpf/ViewModels/Core/NoteViewModel.cs [hunk 2]
```diff
@@ -96,6 +96,20 @@ public bool IsSelected
...
+        /// <summary>
+        /// Property determines if note is being edited, 
+        /// is set to true when double clicking the note
+        /// is set to false when note's textbox looses focus
+        /// </summary>
+        [JsonIgnore]
+        public bool IsOnEditMode
+        {
+            get { return isOnEditMode; }
+            set { isOnEditMode = value; RaisePropertyChanged("IsOnEditMode"); }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/NoteViewModel.cs:110]: Let's use `RaisePropertyChanged(nameof(IsOnEditMode));` instead

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml [hunk 3]
```diff
@@ -31,51 +31,67 @@
...
+            FontFamily="Trebuchet" 
+            FontSize="12"
+            Foreground="#222"
+            AcceptsReturn="True"
+            AcceptsTab="True"
+            Background="#EEE"
+            Padding="10" TextWrapping="Wrap"
+            LostFocus="noteTextBox_LostFocus"
+            PreviewKeyDown="noteTextBox_PreviewKeyDown"
+            IsEnabled="{Binding Path=IsOnEditMode, Mode=OneWay}"
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/NoteView.xaml:48]: @martinstacey Do you mind moving the inline style definition into the central location `DynamoModern.xml`?

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs [hunk 4]
```diff
@@ -15,6 +17,7 @@ namespace Dynamo.Nodes
 {
     public partial class NoteView : IViewModelView<NoteViewModel>
     {
+        private const double MINIMUM_ZOOM_DIRECT_NODE_EDIT = 0.5;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:20]: Would you put some comments for this? Thanks!

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs [hunk 5]
```diff
@@ -84,12 +87,25 @@ void OnNoteTextPreviewMouseDown(object sender, MouseButtonEventArgs e)
...
+            }
+
             // Setup a binding with the edit window's text field
             var dynamoViewModel = ViewModel.WorkspaceViewModel.DynamoViewModel;
             var editWindow = new EditWindow(dynamoViewModel, true)
             {
                 Title = Dynamo.Wpf.Properties.Resources.EditNoteWindowTitle
             };
+
+            editWindow.EditTextBoxPreviewKeyDown += noteTextBox_PreviewKeyDown;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:107]: unsubscribe it somewhere?

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs [hunk 6]
```diff
@@ -152,5 +168,268 @@ private void PrepareZIndex()
...
+        private string BulletTabHandler(string text, int caretIndex)
+        {
+            // Get the line where TAB was typed 
+            var lineNumber = GetLineNumberAtCaretsIndex(text, caretIndex);
+            var line =  GetLineTextAtCaretsIndex(text,caretIndex);
+            var caretAtLine = GetCaretIndexRelativeToLine(text, caretIndex);
+
+            // If TAB was pressed just after a bullet insert it before the bullet
+            // so that you can easily indent
+            var distanceFromBulletToCaret = SPACING_AFTER_BULLET + 1;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:281]: Do we need to define this `1` as const somehow?

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs [hunk 7]
```diff
@@ -152,5 +168,268 @@ private void PrepareZIndex()
...
+            var tabsBeforeBullet = line.Split(BULLETS_CHARS)[0].Count(c => c=='\t');
+            for (int i = 0; i < tabsBeforeBullet; i++)
+            {
+                text = text.Insert(caretIndex, "\t");
+                caretIndex++;
+            }
+            return text.Insert(caretIndex, bulletsInLine.First()+ new String(' ', SPACING_AFTER_BULLET));
+        }
+
+        private bool StringIsEmptyOrTab(string text)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:335]: Hmm by reading the code, seems we are checking if the string input is a combination of spaces or tabs? If that is the case, maybe 
- name this `IsStringSpaceswithTabs`
- use text.All(c=>c==' ' or c=='\t')
Or maybe I read it wrong? Please elaborate

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs [hunk 8]
```diff
@@ -152,5 +168,268 @@ private void PrepareZIndex()
...
+
+        private string TabToSpaceConversion(string text)
+        {
+            return text.Replace("\t", new String(' ', TAB_SPACING_SIZE));
+        }
+
+        private string SpaceToTabConversion(string text)
+        {
+            return text.Replace(new String(' ', TAB_SPACING_SIZE), "\t");
+        }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:432]: I feel a few of these fall into the category of Utility functions which could be shared.. Would you create `StringUtils.cs` under `DynamoUtilities` project and move some of the sprint manipulation functions over?

**Comment by mjkkirschner** [comment]
What does this do to existing notes that might have tabs?

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs (hunk 7 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs]: ok, I agree, `IsStringSpaceswithTabs` is a better name and `String.IsEmpty(text) || text.All(c=>c==' ' or c=='\t')` avoids using a double negative

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs (hunk 6 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:296]: @QilongTang the `SPACING_AFTER_BULLET` is the constant. I think `distanceFromBulletToCaret` doesn't need to be another constant as its always a variable dependant on `SPACING_AFTER_BULLET + 1`, 
It just adds 1 as the caret index is at the end of the character being typed.

Example:
   *_ _ -> two spaces
   *_ _ | -> caretIndex is two spaces + 1 to bullet index.

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs (hunk 5 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:114]: ok

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs (hunk 4 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:25]: ok

File: src/DynamoCoreWpf/ViewModels/Core/NoteViewModel.cs (hunk 2 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/ViewModels/Core/NoteViewModel.cs]: ok

File: src/DynamoCoreWpf/UI/Prompts/EditWindow.xaml.cs (hunk 1 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/UI/Prompts/EditWindow.xaml.cs:50]: ok

**Comment by martinstacey** [comment]
> > What does this do to existing notes that might have tabs?
> 
> Correct me if I am wrong @martinstacey , I think this PR does not affect notes serialized but only affect moving forward from Dynamo 2.13.0, when user type notes certain way using tabs, bullets will be added by Dynamo as interpretation.

@QilongTang  @mjkkirschner 

Yes, so this PR will only affect the user experience, and not the definition of the notes.

- It affects the way you edit a note, (directly through a textbox) 

- It interprets what the user types on the keyboard, to translate it into bullet characters.

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs (hunk 8 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs:447]: ok, @QilongTang, I just added the functions that are applicable to all strings to the `StringUtils.cs` class which are:
`ReplaceLineOfText` , `IsStringSpacesWithTabs` , `BreakTextIntoLines` , `TabToSpaceConversion` , `SpaceToTabConversion`
Ive kept the ones that are exclusively related to bullet point support in `NoteView.xaml.cs` which are:
`BulletDashHandler` , `BulletTabHandler` , `BulletEnterHandler` , `IsCaretRightAfterBullet` , `IsCaretBeforeABullet` , `GetCaretIndexRelativeToLine` , `GetLineNumberAtCaretsIndex` , `GetLineTextAtCaretsIndex` , and `UpdateBulletAccordingToIndentation`
let me know if this is ok

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml (hunk 3 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml:52]: ok addressed in c7dd3c7

File: src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs (hunk 8 above)
Comment (martinstacey) [reply src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs]: addressed in 
28b73d9

**Comment by QilongTang** [comment]
Thanks @martinstacey to address all the comments. The only question I have now is that can you test the current behavior of node with warning and node with note to see how it behaves?

**Comment by martinstacey** [comment]
> Thanks @martinstacey to address all the comments. The only question I have now is that can you test the current behavior of node with warning and node with note to see how it behaves?

Hey @QilongTang, do you mean in the `Pin note to node` command? we are addressing `Pin note to node` in a separate PR as it was a bit more challenging. But that PR is almost ready, I'll add a gif of the warning behavior in that PR, as it is already solved.

**Comment by QilongTang** [comment]
Thank you @martinstacey Then this PR looks good to me. We would like to target this feature to 2.13.0 so will merge when the release branch of 2.12.0 is out

**Comment by Amoursol** [comment]
> Hi @Amoursol, @QilongTang to confirm on both things pending on this PR, editing using `Edit...` both far and close.
> Adding space after bullet
> 

LGTM @martinstacey! (Was on vacation last week - hence the late reply 😊 )

**Review by QilongTang — APPROVED** [review]
LGTM. Will merge once master bumped to 2.13

**Pull request merged**